VMware: Support parsing FORCE-RUN-POST-CUST-SCRIPT

xiaofengw-vmware opened the pull request:
vCloud product wants to run custom script despite of VMware Tools configuration, so we introduce a new flag FORCE-RUN-POST-CUST-SCRIPT. If it is yes, custom script run no matter enable-custom-scripts is true or false.

xiaofengw-vmware commented:
> ```diff
> set_customization_status
> ```

Thanks for your comments. When I follow your comment, I couldn't  pass the flake8 check for below lines:
set_customization_status(
    GuestCustStateEnum.GUESTCUST_STATE_RUNNING,
    GuestCustError.GUESTCUST_ERROR_SCRIPT_DISABLED)
It report that the 3rd lines exceed 79 characters. And it seems I couldn't split it into 2 lines. Do you have any suggestions?

smoser commented:
Can the user disable FORCE-RUN-POST-CUST-SCRIPT in any way?

I can see that some users may be helped by this.  Personally, I do not think should default to allowing arbitrary execution of code by the platform in the guest.  But it definitely should allow the user to disable this behavior.

xiaofengw-vmware commented:
> Can the user disable FORCE-RUN-POST-CUST-SCRIPT in any way?
> 
> I can see that some users may be helped by this. Personally, I do not think cloud-init should, by default, allow arbitrary execution of code by the platform in the guest. If the default behavior is to do so, then it needs to allow the user to override that behavior in user-data.

Thanks Scott. There is no way to disable FORCE-RUN-POST-CUST-SCRIPT. I agree with your that an arbitrary script should be disabled by default.That's the reason why we introduced enable-custom-scripts into the virtual machine configuration file so that guest user could decide whether a script could be executed by vsphere administartor, by default  it is disabled. 
After we introduced it, our vCloud product customer complained that it break their regression and they prefer to enable it by default.  To resolve this, we made this change and vCloud adminstrator will configure FORCE-RUN-POST-CUST-SCRIPT to yes and skip checking enable-custom-scripts. This new API is only exposed to vCloud administrator, other administrators have no change to configure FORCE-RUN-POST-CUST-SCRIPT. How do you think?Thanks.

smoser commented:
@xiaofengw-vmware 

> Thanks Scott. There is no way to disable FORCE-RUN-POST-CUST-SCRIPT. I agree with your that an arbitrary script should be disabled by default.That's the reason why we introduced enable-custom-scripts into the virtual machine configuration file so that guest user could decide whether a script could be executed by vsphere administartor, by default it is disabled.
> After we introduced it, our vCloud product customer complained that it break their regression and they prefer to enable it by default. To resolve this, we made this change and vCloud adminstrator will configure FORCE-RUN-POST-CUST-SCRIPT to yes and skip checking enable-custom-scripts. This new API is only exposed to vCloud administrator, other administrators have no change to configure FORCE-RUN-POST-CUST-SCRIPT. How do you think?Thanks.

The above isn't clear to me.

Cloud-init supports vendor-data, but user-data overrides vendor-data, and can completely disable it.  Thus, if the user (the one who starts the instance) wants to disable the platform (or "vmware administrator"), they can.

If cloud-init is providing a facility for someone other than the user to execute arbitrary code in an instance, then cloud-init needs to provide the user a way to disable that.  If that upsets the vmware administrator then so be it.

If I'm about to start an instance on vmware and this pull request is accepted, can I disable vmware administrator access?

xiaofengw-vmware commented:
> FORCE-RUN-POST-CUST-SCRIPT

Thanks Scott, I understood your concern. I will change this flag to DEFAULT-RUN-POST-CUST-SCRIPT. This flag will only change the default behavior if enable-custom-scripts is absent in VM Tools configuration. If enable-custom-scripts is set , then DEFAULT-RUN-POST-CUST-SCRIPT will be ignored so that VM administrator could decide whether to execute the script. Is this solution ok for you? I will draft the codes and send you for review. Thanks.

smoser commented:
@xiaofengw-vmware,

My request is simple: 
 * user should be able to disable all execution of code that was not inside the image or provided by the user (user-data)
 * this is documented how to do so.

xiaofengw-vmware commented:
> @xiaofengw-vmware,
> 
> My request is simple:
> 
> * user should be able to disable all execution of code that was not inside the image or provided by the user (user-data)
[Xiaofengw]: Yes, user could disable the code by setting in VM Tools configuration file.

> * this is documented how to do so.
[Xiaofengw]: We will document this in our VM Tools administration document. Do you think whether we need to document this in cloud-init? Thanks.

xiaofengw-vmware commented:
> @xiaofengw-vmware,
> 
> My request is simple:
> 
> * user should be able to disable all execution of code that was not inside the image or provided by the user (user-data)
> * this is documented how to do so.

Thank you all for the comments. We have changed our design and introduce a flat called "DEFAULT-RUN-POST-CUST-SCRIPT". This flag only change the default value of "enable-custom-scripts" to yes. So "enable-custom-scripts" will be "yes" in case it is absent. If "enable-custom-scripts" is explicitly set in VM Tools configuration, we still follow its value to enable/disable the script. Please review below pull request. Thanks.
https://github.com/canonical/cloud-init/pull/441
